torch/_dynamo/graph_region_tracker.py

```diff
@@ -0,0 +1,290 @@
+        self.dispatch_table.update(
+            {
+                FakeTensor: _extract_tensor_metadata_for_node_hash,
+                torch.SymInt: lambda x: (_ident, (str(x),)),
+            }
+        )
+        self.fast = True
+
+    def dumps(self, obj: Any) -> bytes:
+        """
+        Pickle an object and return a byte string.
+        """
+        try:
+            self.dump(obj)
+            return self._stream.getvalue()
+        finally:
+            self._stream.seek(0)
+            self._stream.truncate(0)
+
+
+def _extract_tensor_arg(arg: Any) -> Any:
+    if isinstance(arg, Node):
+        return arg.meta.get("example_value")
+    else:
+        return None
+
+
+def _normalize_args(
+    node: Node,
+) -> Tuple[Tuple[str, ...], Tuple[Optional[Any], ...]]:
+    flat_args, _ = tree_flatten(node.args)
+    sorted_kwargs = sorted(node.kwargs.items(), key=lambda x: x[0])
+    sorted_keys = tuple(sorted(node.kwargs.keys()))
+    flat_kwargs, _ = tree_flatten(sorted_kwargs)
+    all_args = flat_args + flat_kwargs
+    return (sorted_keys, tuple(_extract_tensor_arg(arg) for arg in all_args))
+
+
+def get_global_state_key() -> GlobalStateKey:
+    return (
+        torch.is_grad_enabled(),
+        torch.is_inference_mode_enabled(),
+        torch.get_num_threads(),
+        torch._C._get_cublas_allow_fp16_reduced_precision_reduction(),
+        torch._C._get_cublas_allow_bf16_reduced_precision_reduction(),
+        torch.get_default_dtype(),
+        torch.are_deterministic_algorithms_enabled(),
+        torch._C._get_cublas_allow_tf32(),
+        torch.is_deterministic_algorithms_warn_only_enabled(),
+    )
+
+
+# This is typical BFS with the caveat
+# that a node's children need to be explicitly
+# added with the add_children() method
+# The flow is yield a node and check if it's valid for all regions
+# if not valid, discard and continue onto the next node
+class BfsRegionIter:
+    def __init__(self, origin: Node) -> None:
+        self._cur: Tuple[Optional[str], Optional[Node]] = (None, origin)
+        self._queue: Deque[Tuple[Optional[str], Optional[Node]]] = deque()
+
+    @staticmethod
+    def create(origin: Node) -> "BfsRegionIter":
+        it = BfsRegionIter(origin)
+        it.add_children(origin)
+        return it
+
+    def next(self) -> Tuple[Optional[str], Optional[Node]]:
+        ret = self._cur
+        if not self._queue:
+            self._cur = (None, None)
+        else:
+            self._cur = self._queue.popleft()
+        return ret
+
+    def peek(self) -> Tuple[Optional[str], Optional[Node]]:
+        return self._cur
+
+    def add_children(self, node: Node) -> None:
+        arg: Any
+        for arg in node.args:
+            if isinstance(arg, Node):
+                self._append((None, arg))
+
+        key: str
+        kwarg: Any
+        for key, kwarg in node.kwargs.items():
+            if isinstance(kwarg, Node):
+                self._append((key, kwarg))
```

Comment:
This doesn't handle List[Tensor] as input, right? Does that affect the correctness of the algorithm? (Can we add a test for it?)

Reply:
Yeah I'll add tests for this, I'll flatten actually, good catch